keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,344 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The main goal of this enhancement is to authoritatively identify Windows
+pods during the API Server admission time. 
+
+
+## Motivation
+
+Identifying Windows pods during the API Server admission time is crucial to apply
+appropriate security constraints to the pod. Without proper identification, some admission
+plugins may apply unnecessary security constraints to the Windows pods or in the
+worst case, don't apply security constraints at all.
+
+
+### Goals
+
+Admission plugins and/or Admission Webhooks identifying if the incoming pod
+is Windows or not.
+
+
+### Non-Goals
+
+
+- Validating if the Windows pods have appropriate security constraints.
+- Interaction with container runtimes.
+
+## Proposal
+
+
+As mentioned [earlier](#Motivation) identifying the Windows pods during the scheduling phase can be done
+in the following ways:
+- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The problem with using nodeSelector and tolerations is that any unprivileged user/entity can apply the 
+nodeSelector and/or tolerations in the pod spec making it unsecure whereas RuntimeClasses are [recommended](https://kubernetes.io/docs/concepts/containers/runtime-class/#setup) to be
+created by cluster administrator making them authoritative enough to be used.
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+As a Kubernetes cluster administrator, I want appropriate security contexts to be 
+applied to my Windows Pods along with Linux pods
+
+#### Story 2
+
+As a Kubernetes cluster administrator, I want to use my own admission webhook for
+Windows pods.
+
+### Notes/Constraints/Caveats (Optional)
+
+While the `scheduling` field in the RuntimeClass is created to handle scheduling constraints, we have a reserved label
+[`kubernetes.io/os`](https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-os) to identify
+Windows nodes from their Linux counterparts. If the RuntimClass's scheduling field has a nodeSelector  `kubernetes.io/os: 'windows'`,
+we can authoritatively say that the pod is indeed a Windows one during the api-server admission time. Querying the RuntimeClass
```

Comment:
This proposal came up in the context of Pod Security admission. The concern with using scheduling constraints to enforce admission policy is that the scheduler can be bypassed, for example by creating a pod with `NodeName` already set, or issuing a `bind` request manually (with the correct permissions).

To be secure, I'd like the Kubelet to recognize the same definitive source-of-truth that the admission controllers are using. In other words, if the admission decision is based on the nodeSelector, then the Kubelet should also verify the node selector OS, and reject the pod if there's a mismatch.

Reply:
> This proposal came up in the context of Pod Security admission. The concern with using scheduling constraints to enforce admission policy is that the scheduler can be bypassed, for example by creating a pod with NodeName already set, or issuing a bind request manually (with the correct permissions).

That's a good point. We can make kubelet follow the same logic during admission time. As of now, we have logic in kubelet which strips certain security constraints based on OS. I'll add some text which includes changes to kubelet soon.

Reply:
> To be secure, I'd like the Kubelet to recognize the same definitive source-of-truth that the admission controllers are using. In other words, if the admission decision is based on the nodeSelector, then the Kubelet should also verify the node selector OS, and reject the pod if there's a mismatch.

Great idea

Reply:
@derekwaynecarr / @dchen1107 here is another proposed kubelet behavior.
I like the idea of adding in a kubelet check to make sure the current GOOS matches `kubernetes.io/os` nodeSelector (if set for the pod) and fail if they don't match.
Any objections to adding this check?

Reply:
i think that makes sense to add, but i also am fine with not needing the selector at all in favor of a desiredOS attribute.